### ISSUE TITLE
pyarrow on Linux pulls in old version of grpc-cpp

### ISSUE THREAD

**Issue by rolweber** [ContinuumIO/anaconda-issues#12378]
Hello Team Anaconda,

I just noticed that the current pyarrow 3.0.0 installs with a very old grpc-cpp 1.26.0. The reason is that arrow-cpp, a dependency of pyarrow, requires libprotobuf 3.11.2 with an unspecified version of grpc-cpp. Current Anaconda builds of grpc-cpp 1.37.0 require libprotobuf 3.14.0. Therefore, only the old version satisfies the requirements.

One way to resolve this would be to build grpc-cpp for two versions of libprotobuf, the current 3.14.0 and the one required by arrow-cpp, 3.11.2.

available versions of grpc-cpp on linux-64:
```txt
grpc-cpp                      1.26.0      hf484d3e_1  pkgs/main           
grpc-cpp                      1.26.0      hf8bcb03_0  pkgs/main           
grpc-cpp                      1.26.0      hf8bcb03_1  pkgs/main           
grpc-cpp                      1.33.2      h2157cd5_4  pkgs/main           
grpc-cpp                      1.34.1      h2157cd5_4  pkgs/main           
grpc-cpp                      1.35.0      h2157cd5_4  pkgs/main           
grpc-cpp                      1.36.0      h2157cd5_4  pkgs/main           
grpc-cpp                      1.36.1      h2157cd5_4  pkgs/main           
grpc-cpp                      1.36.2      h2157cd5_4  pkgs/main           
grpc-cpp                      1.36.3      h2157cd5_4  pkgs/main           
grpc-cpp                      1.36.4      h2157cd5_4  pkgs/main           
grpc-cpp                      1.37.0      h2157cd5_4  pkgs/main
```

Installation plan for pyarrow:
`conda create --dry-run -n Arrow pyarrow`
```txt
The following NEW packages will be INSTALLED:

  _libgcc_mutex      pkgs/main/linux-64::_libgcc_mutex-0.1-main
  arrow-cpp          pkgs/main/linux-64::arrow-cpp-3.0.0-py39hced866c_0
  blas               pkgs/main/linux-64::blas-1.0-mkl
  boost-cpp          pkgs/main/linux-64::boost-cpp-1.73.0-h27cfd23_11
  brotli             pkgs/main/linux-64::brotli-1.0.9-he6710b0_2
  bzip2              pkgs/main/linux-64::bzip2-1.0.8-h7b6447c_0
  c-ares             pkgs/main/linux-64::c-ares-1.17.1-h27cfd23_0
  ca-certificates    pkgs/main/linux-64::ca-certificates-2021.4.13-h06a4308_1
  certifi            pkgs/main/linux-64::certifi-2020.12.5-py39h06a4308_0
  double-conversion  pkgs/main/linux-64::double-conversion-3.1.5-he6710b0_1
  gflags             pkgs/main/linux-64::gflags-2.2.2-he6710b0_0
  glog               pkgs/main/linux-64::glog-0.4.0-he6710b0_0
  grpc-cpp           pkgs/main/linux-64::grpc-cpp-1.26.0-hf8bcb03_0
  icu                pkgs/main/linux-64::icu-58.2-he6710b0_3
  intel-openmp       pkgs/main/linux-64::intel-openmp-2021.2.0-h06a4308_610
  ld_impl_linux-64   pkgs/main/linux-64::ld_impl_linux-64-2.33.1-h53a641e_7
  libboost           pkgs/main/linux-64::libboost-1.73.0-h3ff78a5_11
  libevent           pkgs/main/linux-64::libevent-2.1.8-h1ba5d50_1
  libffi             pkgs/main/linux-64::libffi-3.3-he6710b0_2
  libgcc-ng          pkgs/main/linux-64::libgcc-ng-9.1.0-hdf63c60_0
  libprotobuf        pkgs/main/linux-64::libprotobuf-3.11.2-hd408876_0
  libstdcxx-ng       pkgs/main/linux-64::libstdcxx-ng-9.1.0-hdf63c60_0
  libthrift          pkgs/main/linux-64::libthrift-0.13.0-hfb8234f_6
  lz4-c              pkgs/main/linux-64::lz4-c-1.9.3-h2531618_0
  mkl                pkgs/main/linux-64::mkl-2021.2.0-h06a4308_296
  mkl-service        pkgs/main/linux-64::mkl-service-2.3.0-py39h27cfd23_1
  mkl_fft            pkgs/main/linux-64::mkl_fft-1.3.0-py39h42c9631_2
  mkl_random         pkgs/main/linux-64::mkl_random-1.2.1-py39ha9443f7_2
  ncurses            pkgs/main/linux-64::ncurses-6.2-he6710b0_1
  numpy              pkgs/main/linux-64::numpy-1.20.1-py39h93e21f0_0
  numpy-base         pkgs/main/linux-64::numpy-base-1.20.1-py39h7d8b39e_0
  openssl            pkgs/main/linux-64::openssl-1.1.1k-h27cfd23_0
  orc                pkgs/main/linux-64::orc-1.6.5-h973521d_1
  pandas             pkgs/main/linux-64::pandas-1.2.4-py39h2531618_0
  pip                pkgs/main/linux-64::pip-21.0.1-py39h06a4308_0
  pyarrow            pkgs/main/linux-64::pyarrow-3.0.0-py39he0739d4_3
  python             pkgs/main/linux-64::python-3.9.4-hdb3f193_0
  python-dateutil    pkgs/main/noarch::python-dateutil-2.8.1-pyhd3eb1b0_0
  pytz               pkgs/main/noarch::pytz-2021.1-pyhd3eb1b0_0
  re2                pkgs/main/linux-64::re2-2020.11.01-h2531618_1
  readline           pkgs/main/linux-64::readline-8.1-h27cfd23_0
  setuptools         pkgs/main/linux-64::setuptools-52.0.0-py39h06a4308_0
  six                pkgs/main/linux-64::six-1.15.0-py39h06a4308_0
  snappy             pkgs/main/linux-64::snappy-1.1.8-he6710b0_0
  sqlite             pkgs/main/linux-64::sqlite-3.35.4-hdfb4753_0
  tk                 pkgs/main/linux-64::tk-8.6.10-hbc83047_0
  tzdata             pkgs/main/noarch::tzdata-2020f-h52ac0ba_0
  uriparser          pkgs/main/linux-64::uriparser-0.9.3-he6710b0_1
  utf8proc           pkgs/main/linux-64::utf8proc-2.6.1-h27cfd23_0
  wheel              pkgs/main/noarch::wheel-0.36.2-pyhd3eb1b0_0
  xz                 pkgs/main/linux-64::xz-5.2.5-h7b6447c_0
  zlib               pkgs/main/linux-64::zlib-1.2.11-h7b6447c_3
  zstd               pkgs/main/linux-64::zstd-1.4.5-h9ceee32_0
```

**Comment by rolweber** [comment]
Another solution could be to drop the run-time dependency altogether, in favor of linking arrow-cpp statically against grpc-cpp.

**Comment by rolweber** [comment]
The situation is more intricate than I imagined when I opened this issue. The [Arrow Flight protocol](https://arrow.apache.org/docs/format/Flight.html) is not defined on the wire level, but as a layer on top of grpc, therefore protobuf. And protobuf is incompatible across minor version... in some cases even across fixlevels: https://abi-laboratory.pro/?view=timeline&l=protobuf

For the Arrow Flight protocol to work, client and server need to use compatible versions of protobuf. If Anaconda main has arrow-cpp statically linked against a specific protobuf version, that will work only with some servers, but not with others. If it's dynamically linked, dependency conflicts will arise if other packages like Tensorflow require a particular version of protobuf.

I don't even know if it's possible to select the protobuf version for grpc at installation time with dynamic linking. Probably not, if the grpc protocol definitions need to be processed at build time with the specific version of protobuf that will be used at runtime. In that case, it would make more sense to link statically and to avoid a runtime dependency on grpc-cpp and protobuf altogether.

Our in-house teams are investigating how to handle our dependencies on protobuf versions for the Arrow Flight protocol. We might have to build our own packages for Arrow, regardless of what Anaconda main provides.
